server/schedulers/fail_over.go

```diff
@@ -0,0 +1,192 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedulers
+
+import (
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/schedule"
+	log "github.com/sirupsen/logrus"
+)
+
+func init() {
+	schedule.RegisterScheduler("fail-over", func(limiter *schedule.Limiter, args []string) (schedule.Scheduler, error) {
+		return newFailOverScheduler(limiter), nil
+	})
+}
+
+type failOverScheduler struct {
+	*baseScheduler
+	leaderSelector schedule.Selector
+}
+
+func newFailOverScheduler(limiter *schedule.Limiter) schedule.Scheduler {
+	filters := []schedule.Filter{
+		schedule.NewStateFilter(),
+		schedule.NewHealthFilter(),
+		schedule.NewSnapshotCountFilter(),
+		schedule.NewStorageThresholdFilter(),
+		schedule.NewPendingPeerCountFilter(),
+	}
+	base := newBaseScheduler(limiter)
+	return &failOverScheduler{
+		baseScheduler:  base,
+		leaderSelector: schedule.NewBalanceSelector(core.RegionKind, filters),
+	}
+}
+
+func (f *failOverScheduler) GetName() string {
+	return "fail-over-scheduler"
+}
+
+func (f *failOverScheduler) GetType() string {
+	return "fail-over"
+}
+
+func (f *failOverScheduler) IsScheduleAllowed(cluster schedule.Cluster) bool {
+	return f.limiter.OperatorCount(schedule.OpReplica) < cluster.GetReplicaScheduleLimit()
+}
+
+func (f *failOverScheduler) Schedule(cluster schedule.Cluster, opInfluence schedule.OpInfluence) []*schedule.Operator {
+	stores := cluster.GetStores()
+	for _, store := range stores {
+		if store.DownTime() > cluster.GetMaxStoreDownTime() {
+			op := f.handleDownStore(cluster, store)
+			if op != nil {
+				op.SetPriorityLevel(core.HighPriority)
+				return []*schedule.Operator{op}
+			}
+		}
+		if store.IsOffline() {
+			op := f.handleOfflineStore(cluster, store)
+			if op != nil {
+				op.SetPriorityLevel(core.HighPriority)
+				return []*schedule.Operator{op}
+			}
+		}
+		if store.IsLowSpace() {
+			op := f.handleLowSpaceStore(cluster, store)
+			if op != nil {
+				op.SetPriorityLevel(core.HighPriority)
+				return []*schedule.Operator{op}
+			}
+		}
+	}
+	return nil
+}
+
+func (f *failOverScheduler) handleDownStore(cluster schedule.Cluster, store *core.StoreInfo) *schedule.Operator {
+	region := cluster.RandLeaderRegion(store.GetId())
+	if region != nil {
+		log.Warnf("[region %d] Down Store [%v] have leader region, Maybe have network problem.", region.GetId(), store)
+		schedulerCounter.WithLabelValues(f.GetName(), "down_leader").Inc()
+		return nil
+	}
+	region = cluster.RandFollowerRegion(store.GetId())
+	if region == nil {
+		return nil
+	}
+	if region.HealthPeerCount() < cluster.GetMaxReplicas()/2+1 {
+		log.Errorf("[region %d] region unhealth: %s", region.GetId(), region)
+		schedulerCounter.WithLabelValues(f.GetName(), "unhealth").Inc()
+		return nil
+	}
+
+	peer := region.GetStorePeer(store.GetId())
+	if peer == nil {
+		return nil
+	}
+	downPeer := region.GetDownPeer(peer.GetId())
```

Comment:
is it possible that peer down but store is not down?

Reply:
No one guarantees this, down peer reported by TiKV, and down store judge by PD.